### PR TITLE
Automatically disable the HTTP port when client certs are required

### PR DESCRIPTION
**Pull request by cescoffier**
This is particularly important when mTLS is used.

### CODE REVIEW

**Review by gsmet — APPROVED** [review]

File: docs/src/main/asciidoc/security-authentication-mechanisms.adoc
Comment (sberyozkin) [inline docs/src/main/asciidoc/security-authentication-mechanisms.adoc:196]: @cescoffier In this context it makes no difference because Quarkus Security enforces it with its own constraints, the `authenticated` policy will fail.

**Review by sberyozkin — DISMISSED** [review]
IMHO if `REQUIRED` is set then no plain port should be open by default, so the proposal is to fail the build if insecure requests are not disabled but doc that users can open it if they really need it

**Comment by sberyozkin** [comment]
@cescoffier 

>the proposal is to fail the build if insecure requests are not disabled but doc that users can open it if they really need it

That was incomplete, having `REQUEST` would allow such requests and enable a semi secured transport level protection.

I honestly don't think disabling insecure requests on the main router with `REQUIRED` can break anyone's code, as `REQUIRED` is a very purposeful mode. 

I can open a PR if you'd like me to.

File: docs/src/main/asciidoc/security-authentication-mechanisms.adoc
Comment (michalvavrik) [reply docs/src/main/asciidoc/security-authentication-mechanisms.adoc:196]: Are you sure using `quarkus.http.insecure-requests=disabled ` isn't more effective than HTTP security policy when it comes to disabling HTTP protocol? I think it means Vert.x will not pass any insecure requests to route handlers so there will be way less of processing. The `quarkus.http.insecure-requests=disabled ` configuration property should have positive impact.

Comment (sberyozkin) [reply docs/src/main/asciidoc/security-authentication-mechanisms.adoc:196]: @michalvavrik May be it is is but in this example it is unnecessary, where Quarkus security is already configured.

I'd consider documenting it here as an alternative, IMHO that would better IMHO

**Comment by sberyozkin** [comment]
My only doubt is whether `IMHO if REQUIRED is set then no plain port should be open by default` is correct or not. May be it sounds too restrictive as users may want to limit MTLS to specific paths, but it appears to me a mainstream MTLS case is to ensure the client is authenticated at the transport level, so having to disable the insecure port seems like an extra action too many in this case. It would be fairer if users who wanted to relax this all around MTLS protection would open the insecure port if they want to...
@cescoffier @michalvavrik, does it sound reasonable ? I think we should try to be more opinionated in this regard, if MTLS is REQUIRED, 100% HTTPS is enforced, that is not too drastic, if it appears to be the case to some users, just open the insecure port, instead of users who want to do 100% HTTPS having to do it :-)

File: docs/src/main/asciidoc/security-authentication-mechanisms.adoc
Comment (michalvavrik) [reply docs/src/main/asciidoc/security-authentication-mechanisms.adoc:196]: Provided app use `io.quarkus.security.runtime.X509IdentityProvider` and don't replace it with something more complex, I think it results in a same thing, yeah. If it's documented separately with context user will have all information.

**Comment by michalvavrik** [comment]
> My only doubt is whether IMHO if REQUIRED is set then no plain port should be open by default is correct or not. May be it sounds too restrictive as users may want to limit MTLS to specific paths, but it appears to me a mainstream MTLS case is to ensure the client is authenticated at the transport level, so having to disable the insecure port seems like an extra action too many in this case. It would be fairer if users who wanted to relax this all around MTLS protection would open the insecure port if they want to...
@cescoffier @michalvavrik @gsmet , does it sound reasonable ? I think we should try to be more opinionated in this regard, if MTLS is REQUIRED, 100% HTTPS is enforced, that is not too drastic, if it appears to be the case to some users, just open the insecure port, instead of users who want to do 100% HTTPS having to do it :-)

I think it sounds reasonable. It's also potentially breaking change, FWIW +1

**Comment by cescoffier** [comment]
@sberyozkin There is a problem with this solution. Both REQUIRED and REDIRECTED make sense. It's up to the user to decide. 

I agree that REQUIRED is probably better, but it is a strong opinion. So, we would need to log something explaining that we took this decision because the user didn't set any value.

**Comment by cescoffier** [comment]
@sberyozkin is there an easy way to know if mTLS is configured? I was wondering if looking if `tlsClientAuth` is set to `REQUIRED`. However, I do not believe it capture everything.

**Comment by sberyozkin** [comment]
@cescoffier Right, at build time it is checked if it is set to `REQUIRED` or `REQUEST` and if yes then it is assumed it is enabled.

I've dismissed my review, even though I feel it is worth be stricter by default, when `REQUIRED` is set, I don't understand all the consequences of disabling insecure ports by default.
Also, when the Quarkus Security is used to enforce the authentication alongside `REQUIRED`, insecure HTTP requests will be rejected, but thinking more about it, disabling them earlier at the transport level may actually be more performant.

Please fix the conflict, sorry it took me a while to think about it

**Review by sberyozkin — APPROVED** [review]

**Comment by cescoffier** [comment]
@sberyozkin Can you review, I automatically disable the plain server.

**Comment by sberyozkin** [comment]
@cescoffier 

> I automatically disable the plain server.

That is the Quarkus opinionated way :-) :+1: 
Worth trying for 3.7.0.CR1 and see if there will be some concerns, may be only start with the warning, let me comment there

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/options/HttpServerOptionsUtils.java
Comment (sberyozkin) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/options/HttpServerOptionsUtils.java:440]: @cescoffier 

Should we start with `warn` in cases when the users decided to enable it ? I'm thinking if they decided to do it then there might be a good reason for it, one case I can imagine, which may be rare enough, but one can restrict MTLS Required to specific paths, 

https://quarkus.io/guides/security-authentication-mechanisms#path-specific-authentication-mechanisms

for example:
```
quarkus.http.auth.permission.mtls.paths=/mtls-required
quarkus.http.auth.permission.mtls.policy=authenticated
quarkus.http.auth.permission.mtls.auth-mechanism=mtls
```

and do something else for other paths. I have to admit, I can't really imagine why, in prod, one would have a semi-secured application like that, where the client certificate is enforced on some paths, and then anyone can access that app over plain HTTP on some other paths.

But may be `warn` would be a bit softer, and then if users won't push back a lot on this warning in 3.7.0.CR1/3.7.0, we can totally enforce as you propose with this exception in 3.8.0, what do you think ?

Comment (sberyozkin) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/options/HttpServerOptionsUtils.java:440]: It should fix the failing tests

**Comment by sberyozkin** [comment]
I can propose a migration note once this one is merged, even though technically it is a breaking change, as suggested above, I have high hopes very few users if any will notice it

**Comment by cescoffier** [comment]
The CI issue is most probably related.

**Review by sberyozkin — APPROVED** [review]

**Comment by sberyozkin** [comment]
@cescoffier FYI, I prototypes https://github.com/quarkusio/quarkus/wiki/Migration-Guide-3.7#insecure-http-port-is-disabled-when-mtls-client-authentication-is-required.

Please do not hesitate to change/fix it as you prefer, it would be fine. Thanks

**Comment by sberyozkin** [comment]
@cescoffier This one is a new test, which got into the build after the rebase :-), I can push an update to it to have it accepting insecure requests or the test can validate that the HTTP connection is closed to verify the secure port is effectively disabled - how to correctly check it though ?

**Comment by cescoffier** [comment]
@sberyozkin Connection refused is exactly what we want in this case. Could you adapt the test to check for that exception? Otherwise, enable the the "plain" HTTP server.

**Comment by cescoffier** [comment]
There are a few import * statement to cleanup.

**Comment by sberyozkin** [comment]
These test failure are not related

**Pull request merged**
